src/dirs.rs

```diff
@@ -9,16 +9,23 @@ pub fn local_path_and_canonical_url(
     url: &str,
     cargo_home: Option<&std::path::Path>,
 ) -> Result<(std::path::PathBuf, String), Error> {
-    let (dir_name, canonical_url) = url_to_local_dir(url)?;
-
     let mut path = match cargo_home {
         Some(path) => path.to_owned(),
         None => home::cargo_home()?,
     };
 
     path.push("registry");
     path.push("index");
-    path.push(dir_name);
+
+    let (dir_name, canonical) = url_to_local_dir(url, true)?;
```

Comment:
@frewsxcv This check here doesn't quite work. There could be a case where both directories exist but only one of them has the desired crates populated. To be able to implement a fallback, both paths would need to be tracked on structs like `SparseIndex` so that the crate paths could be checked in either case. Curious on your thoughts there.

Reply:
That's true. When seeing it I thought it was a good way of dealing with those that are transitioning, preferring the new hash location.
Doing so is good enough, I don't think there is a need for more complexity for the mere chance of hitting the cache more often. Cargo probably doesn't do that either.